R/utils-code-dependency.R

```diff
@@ -0,0 +1,500 @@
+          # 'options(prompt = ">")' is put in a one item
+          # and '# @effect ADLB' is the first element of the next item.
+          # This is tackled in B.
+
+
+          if (!com_cond[1] & sum(com_cond) > 0) {
+            # A.
+            x[x$token == "SYMBOL", "text"]
+          } else if (com_cond[1] & sum(com_cond[-1]) > 0) {
+            # B.
+            x <- x[-1, ]
+            x[x$token == "SYMBOL", "text"]
+          }
+        }
+      )
+    )
+
+  commented_calls <- vapply(pd,
+    function(x) any(x$token == "COMMENT" & grepl("@effect", x$text)),
+    FUN.VALUE = logical(1)
+  )
+
+  symbol_effects_lines <-
+    unlist(
+      lapply(
+        symbol_effects_names,
+        function(x) {
+          code <- return_code(x, pd = pd, occur = occur, cooccur = cooccur)
+          # QUESTION: SHOULD cooccur BE TRIMMED like it happens in return_code()?
+          if (is.null(code)) {
+            # Extract lines for objects that were used, but never created.
+            # Some objects like 'iris' or 'mtcars' are pre-assigned in the session.
+            # Below is just used for comments with @effect.
+            # if (!object %in% names(occur)) {
+            intersect(which(detect_symbol(x, pd)), which(commented_calls))
+            # }
+          } else {
+            code
+          }
+        }
+      )
+    )
+
+  # When commet_id is the highest id in the item - take previous item.
+  side_effects_names <-
+    unlist(
+      lapply(
+        pd,
+        function(x) {
+          com_cond <-
+            x$token == "COMMENT" & grepl("@effect", x$text) & grepl(paste0("[\\s]*", object, "[\\s$]*"), x$text)
+
+          # Work out the situation when comment id is the highest id in the item.
+          # For calls like 'options(prompt = ">") # @effect ADLB',
+          # 'options(prompt = ">")' is put in a one item
+          # and '# @effect ADLB' is the first element of the next item.
+
+          com_cond[1]
+        }
+      )
+    )
+
+  side_effects_lines <- which(side_effects_names) - 1
+
+  sort(unique(c(symbol_effects_lines, side_effects_lines)))
+}
+
+#' Return the lines of code (with side-effects) needed to reproduce the object.
+#' @return `character` vector of elements of `parsed_code` calls that were required to build the side-effects and
+#' influencing objects having and impact on the `object`
+#'
+#' @param qenv `qenv` object
+#' @param names `character` with object names
+#' @keywords internal
+get_code_dependency <- function(qenv, names) {
+  filtered_cd <- Filter(function(x) length(x) > 1, qenv@code_dependency)
+  if (identical(filtered_cd, list())) {
+    message(
+      "Code dependency is supported only for the code provided as a character in new_env()/eval_code()."
+    )
+    return(NULL)
+  }
+
+  if (!all(names %in% ls(qenv@env))) {
+    warning(
+      "Objects not found in 'qenv' environment: ",
+      paste(names[!(names %in% ls(qenv@env))], collapse = ", ")
+    )
+  }
```

Comment:
I allow the function to work further, because if someone asks for 3 objects and 1 of them does not exist, you at least get the code for other two. Maybe it's better if we put error in here